File: 2024/D01/d01.js
/*
Advent Of Code 2024
Day 1: Historian Hysteria part 1 & 2

https://adventofcode.com/2024/day/1
*/

const fs = require('fs');
const path = require('path');

const YEAR = '2024';
const DAY = '01';

function elapsedTime(name, fct, input) {
  const startTime = performance.now();
  const result = fct(input);
  const elapsed = performance.now() - startTime;
  console.log(
    `${name} time: ${Math.floor(elapsed / 60000)}:${Math.floor((elapsed % 60000) / 1000)}:${Math.floor(elapsed % 1000)}`
  );
  console.log(`${name} result: ${result}`);
}

function readInput(filename) {
  const readRawInput = () => {
    const data = fs.readFileSync(path.join(`${YEAR}`, `D${DAY}`, filename));
    const input = data
      .toString()
      .split('\n')
      .filter((l) => l.length > 0)
      .map((l) => l.split('   ').map((v) => parseInt(v)));
    return input;
  };
  try {
    return readRawInput();
  } catch (err) {
    console.error(err);
  }
}

function part1(input) {
  // reduce the input in 2 arrays of int
  const [arr1, arr2] = input.reduce(
    (arr, v) => {
      arr[0].push(v[0]);
      arr[1].push(v[1]);
      return arr;
    },
    [[], []]
  );

  // sort them
  arr1.sort();
  arr2.sort();

  // reduce arr1 and calculate distance
  const result = arr1.reduce((r, v, index) => r + Math.abs(v - arr2[index]), 0);
  return result;
}

function part2(input) {
  // reduce the input in 2 arrays of int
  const [arr1, arr2] = input.reduce(
    (arr, v) => {
      arr[0].push(v[0]);
      arr[1].push(v[1]);
      return arr;
    },
    [[], []]
  );

  // reduce arr1 and check number of occurrence in arr2
  const result = arr1.reduce((r, v) => r + v * arr2.filter((e) => e === v).length, 0);
  return result;
}

//const input = readInput(`d${DAY}-sample.txt`);
const input = readInput(`d${DAY}-input.txt`);
elapsedTime('Part 1', part1, input);
elapsedTime('Part 2', part2, input);

File: 2024/D01/d01.py
"""
Advent Of Code 2024
Day 1: Historian Hysteria part 1 & 2

https://adventofcode.com/2024/day/1
"""
from time import perf_counter as pfc
import os

YEAR = "2024"
DAY = "01"

def readInput(filename):
  with open(os.path.join(f"{YEAR}", f"D{DAY}", filename)) as f:
    lines = f.readlines()
  input = ((int(e[0]), int(e[1])) for e in (e.split("   ") for e in (d.strip('\n') for d in lines)))
  return tuple(input)

def elapsedTime(name, fct, input) :
  startTime = pfc()
  result = fct(input)
  print(f"${name} time: {pfc() - startTime:.4} s")
  print(f"${name} result: {result}")


def part1(input):
  tuple1, tuple2  = zip(*input)
  
  # sort the 2 array
  sorted_tuple1 = sorted(tuple1)
  sorted_tuple2 = sorted(tuple2)

  # iterate through 1st tuple and calculate distance 
  result = 0
  for index, e1 in enumerate(sorted_tuple1):
    result += abs(e1- sorted_tuple2[index])

  return result

def part2(input): 
  tuple1, tuple2  = zip(*input)
  # generate a tuple of occurrence count
  occurrences = (v*tuple2.count(v) for v in tuple1)
  return sum(occurrences)

#input = readInput(f"d{DAY}-sample.txt")
input = readInput(f"d{DAY}-input.txt")
elapsedTime("Part 1",part1, input)
elapsedTime("Part 2",part2, input)
